src/v/iceberg/rest_client/catalog_client.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "http/client.h"
+#include "http/request_builder.h"
+#include "iceberg/rest_client/retry_policy.h"
+#include "iceberg/rest_client/types.h"
+#include "json/document.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+namespace iceberg::rest_client {
+
+// helper intended to catch JSON parse errors when parsing an iobuf to json
+// document before passing the document to more specific parsers
+expected<json::Document> parse_json(iobuf&& raw_response);
+
+// The catalog client enables high level operations against a catalog rest api.
+// It also ensures that a token is available before making calls. A client is
+// lightweight and expected to be created for single operations or closely
+// related operations.
+class catalog_client {
+public:
+    /// \brief Construct a catalog client
+    /// \param client an object implementing the abstract client interface which
+    /// can perform an http request
+    /// \param endpoint the rest catalog server hostname to connect to,
+    /// including scheme
+    /// \param credentials credentials to be used to acquire oauth token from
+    /// catalog server
+    /// \param base_url an optional path to be prefixed to all API calls made by
+    /// this client
+    /// \param prefix an optional prefix to be added after the API version,
+    /// applied to all calls made by this client
+    /// \param api_version pi version used to construct URLs, defaults to v1
+    /// \param token an optional oauth token which will be used if valid. If
+    /// expired, a new one will be acquired
+    /// \param retry_policy a retry policy
+    /// used to determine how failing calls will be retried
+    catalog_client(
+      http::abstract_client& client,
```

Comment:
I don't know the expected usage here, but seeing a bunch of unowned in a class like this gives me a bit of pause - who is going to own the memory of these objects?